src/compiler/sys.ts

```diff
@@ -294,6 +294,50 @@ namespace ts {
                 };
             }
 
+            function createWatchedFileSet() {
+                const watchedDirectories: { [path: string]: FileWatcher } = {};
+                const watchedFiles: { [fileName: string]: (fileName: string, removed?: boolean) => void; } = {};
+
+                function addFile(fileName: string, callback: (fileName: string, removed?: boolean) => void): WatchedFile {
+                    const file: WatchedFile = { fileName, callback };
+                    const watchedPaths = Object.keys(watchedDirectories);
+                    // Try to find parent paths that are already watched. If found, don't add directory watchers
+                    const watchedParentPaths = watchedPaths.filter(path => fileName.indexOf(path) === 0);
+                    // If adding new watchers, try to find children paths that are already watched. If found, close them. 
+                    if (watchedParentPaths.length === 0) {
+                        const pathToWatch = ts.getDirectoryPath(fileName);
+                        for (const watchedPath in watchedDirectories) {

```

Comment:
You are right, I was trying to use the recursive `fs.watch` but forgot to specify the `{recursive: true}` option. 

Though on a second thought, this logic may be too complex when I need to close a unused directory watcher later. For non-recursive watchers I can just check if the direct parent folder has any files in the watching list, while for recursive watchers it's more complex. It might make more sense to check to the non-recursive version.